### DISCUSSION TITLE
Sun, Moon, Earth positions and Dark/Full moon formulars/calculations for computer game OpenXcom

### DISCUSSION THREAD

**Discussion by SkybuckFlying** [Stellarium/stellarium#2403] [Q&A]
Hello,

For this remake of the classic ms-dos computer game x-com, done in OpenXcom github project it would be very usefull if there was a 
"universe class" which would maybe link to a "earth class, moon class, sun class" and where their positions are calculated.

The game already has some calculations for crash sites and base sites and such... but for the lighting engine it would be usefull to know the sun position, moon position, earth position and crash site position.... and finally earth+crash site position to create a universe position to understand if the site is in the light of the sun or moon.

So basically the idea is to traverse/cast rays from tiles/voxels towards either the sun or moon and to then compute a shade value from 0 to 15, 0 = dark, 15 = fully lit.

It would also be very usefull if these sun,earth,moon positions/calculations were done based on time/day and such.

Here I discuss these ideas further:

https://openxcom.org/forum/index.php/topic,10497.0.html

The game already has some math but I am not sure how realistic this is, so far it only compute the sun direction:

(I am hoping that one of you programmers of Stellarium is interested in shortly joining this development effort, and maybe supply as with some information, links to information/formulas/calculations, or some guidance/pointers where to find this code/math in stellarium or even better, perhaps supply some isolated formulas/code that can be used for the OpenXcom game which is here:

forum:
https://openxcom.org/forum/
git:
https://github.com/OpenXcom/OpenXcom
)

Please let me know your thoughts on this matter, I am very curious, I am also myself trying to implement better light algorithm and calculation for this game... also voxel traversal algorithm tile traversal algorithm using fast voxel traversal from woo amanatides.

```
/**
 * Get position of sun from point on globe
 * @param lon longitude of position
 * @param lat latitude of position
 * @return position of sun
 */
Cord Globe::getSunDirection(double lon, double lat) const
{
	const double curTime = _game->getSavedGame()->getTime()->getDaylight();
	const double rot = curTime * 2*M_PI;
	double sun;

	if (Options::globeSeasons)
	{
		const int MonthDays1[] = {0, 31, 59, 90, 120, 151, 181, 212, 243, 273, 304, 334, 365};
		const int MonthDays2[] = {0, 31, 60, 91, 121, 152, 182, 213, 244, 274, 305, 335, 366};

		int year=_game->getSavedGame()->getTime()->getYear();
		int month=_game->getSavedGame()->getTime()->getMonth()-1;
		int day=_game->getSavedGame()->getTime()->getDay()-1;

		double tm = (double)(( _game->getSavedGame()->getTime()->getHour() * 60
			+ _game->getSavedGame()->getTime()->getMinute() ) * 60
			+ _game->getSavedGame()->getTime()->getSecond() ) / 86400; //day fraction is also taken into account

		double CurDay;
		if (year%4 == 0 && !(year%100 == 0 && year%400 != 0))
			CurDay = (MonthDays2[month] + day + tm )/366 - 0.219; //spring equinox (start of astronomic year)
		else
			CurDay = (MonthDays1[month] + day + tm )/365 - 0.219;
		if (CurDay<0) CurDay += 1.;

		sun = -0.261 * sin(CurDay*2*M_PI);
	}
	else
		sun = 0;

	Cord sun_direction(cos(rot+lon), sin(rot+lon)*-sin(lat), sin(rot+lon)*cos(lat));

	Cord pole(0, cos(lat), sin(lat));

	if (sun>0)
		 sun_direction *= 1. - sun;
	else
		 sun_direction *= 1. + sun;

	pole *= sun;
	sun_direction += pole;
	double norm = sun_direction.norm();
	//norm should be always greater than 0
	norm = 1./norm;
	sun_direction *= norm;
	return sun_direction;
}
```

Bye,
  Skybuck.

**Comment by gzotti** [comment]
Sorry to disappoint you, but I am fully busy and cannot divert into yet another sparetime project. I don't think Stellarium is the easiest project to reverse-engineer for your purpose, there should be enough open code out there.
